bluemira/codes/process/_solver.py

```diff
@@ -0,0 +1,236 @@
+            files. Default is current working directory.
+        * template_in_dat: str
+            The path to a template IN.DAT file. The default is to use
+            one provided by bluemira.
+        * problem_settings: Mapping[str, float]
+            Any PROCESS parameters that do not correspond to a bluemira
+            parameter.
+
+    Notes
+    -----
+    This solver has several run modes:
+
+    * run: Run PROCESS to generate a radial build.
+        Creates a new input file from the given template IN.DAT, which
+        is modified with bluemira parameters that are mapped with
+        :code:`send = True`.
+    * runinput: Run PROCESS with an unmodified template IN.DAT.
+        The template IN.DAT is not modified with bluemira parameters.
+        This is equivalent to all bluemira parameters mappings having
+        :code:`send = False`.
+    * read: Load the radial build from a PROCESS MFILE.DAT.
+        Loads only the parameters with :code:`send = True`.
+        A file named 'MFILE.DAT' must exist within 'run_directory'.
+    * readall: Load the radial build from a PROCESS MFILE.DAT.
+        Loads all mappable parameters from the PROCESS file.
+        A file named 'MFILE.DAT' must exist within 'run_directory'.
+    * mock: Load bluemira parameters directly from a JSON file in the
+        run directory. This does not run PROCESS.
+    * none: Does nothing.
+        PROCESS is not run and parameters are not updated. This is
+        useful loading results form previous runs of bluemira, where
+        overwriting data with PROCESS outputs would be undesirable.
+    """
+
+    name = PROCESS_NAME
+    setup_cls = Setup
+    run_cls = Run
+    teardown_cls = Teardown
+    run_mode_cls = RunMode
+
+    def __init__(self, params: ParameterFrame, build_config: BuildConfig):
+        # Init task objects on execution so parameters can be edited
+        # between separate 'execute' calls.
+        self._setup: Union[Setup, None] = None
+        self._run: Union[Run, None] = None
+        self._teardown: Union[Teardown, None] = None
+
+        self.params = params
+        add_mapping(PROCESS_NAME, self.params, process_mappings)
+
+        _build_config = copy.deepcopy(build_config)
+        self.binary = _build_config.pop("binary", PROCESS_BINARY)
+        self.run_directory = _build_config.pop("run_dir", os.getcwd())
+        self.read_directory = _build_config.pop("read_dir", os.getcwd())
+        self.template_in_dat = _build_config.pop("template_in_dat", DEFAULT_INDAT)
+        self.problem_settings = _build_config.pop("problem_settings", {})
+        self.in_dat_path = _build_config.pop(
+            "in_dat_path", os.path.join(self.run_directory, "IN.DAT")
+        )
+        if len(_build_config) > 0:
+            bluemira_warn(
+                f"'{self.name}' solver received unknown build_config arguments: "
+                f"'{''', '''.join(_build_config.keys())}'."
```

Comment:
Extra quotes seemingly for no reason? This produces the same result for me:
```suggestion
                f"'{', '.join(_build_config.keys())}'."
```

Reply:
How about this:

```python
In [6]: d = {"a":1, "b":2, "c":3}

In [7]: block = "', '"

In [8]: f"'{block.join(d.keys())}'."
Out[8]: "'a', 'b', 'c'."

```

Reply:
Done 👍. 

Another cool way of doing this I came across: `str(list(d.keys()))[1:-1]`. Less easy to read than your solution though,